base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -1159,6 +1180,119 @@ where
         Ok((tx_id, tx, fee, utxos_total_value))
     }
 
+    async fn fetch_outputs_from_node(
+        &mut self,
+        hashes: Vec<HashOutput>,
+    ) -> Result<Vec<TransactionOutput>, OutputManagerError> {
+        // lets get the output from the blockchain
+        let req = FetchMatchingUtxos { output_hashes: hashes };
+        let results: Vec<TransactionOutput> = self
+            .resources
+            .connectivity
+            .obtain_base_node_wallet_rpc_client()
+            .await
+            .ok_or_else(|| {
+                OutputManagerError::InvalidResponseError("Could not connect to base node rpc client".to_string())
+            })?
+            .fetch_matching_utxos(req)
+            .await?
+            .outputs
+            .into_iter()
+            .filter_map(|o| match o.try_into() {
+                Ok(output) => Some(output),
+                _ => None,
+            })
+            .collect();
+        Ok(results)
+    }
+
+    pub async fn create_claim_sha_atomic_swap_transaction(
+        &mut self,
+        output: TransactionOutput,
+        pre_image: PublicKey,
+        fee_per_gram: MicroTari,
+    ) -> Result<(u64, MicroTari, MicroTari, Transaction), OutputManagerError> {
+        let spending_key = PrivateKey::from_bytes(
+            CommsPublicKey::shared_secret(
+                self.node_identity.as_ref().secret_key(),
+                &output.sender_offset_public_key,
+            )
+            .as_bytes(),
+        )?;
+        let rewind_key = PrivateKey::from_bytes(&hash_secret_key(&spending_key))?;
+        let blinding_key = PrivateKey::from_bytes(&hash_secret_key(&rewind_key))?;
+        let rewound =
+            output.full_rewind_range_proof(&self.resources.factories.range_proof, &rewind_key, &blinding_key)?;
+
+        let rewound_output = UnblindedOutput::new(
+            rewound.committed_value,
+            rewound.blinding_factor.clone(),
```

Comment:
Since you are rewinding anyway, why set the blinding factor to a DH key? Can't the sender use any blinding factor?

Reply:
Well, technically we could. For now, I kept its workings the same as the current one-sided implementation.  But from a technical standpoint, it makes no difference if the blinding factor is a DH key or it is random and only the rewinding keys used is a DH key.

Reply:
if you don't, you risk exposing every swap into and out of tari. A person interested in a specific pub key could easily brute force all the htlc's